### PR TITLE
Normalize associated types in paths in expressions

### PR DESCRIPTION
**Pull request by lowr**
Part of #14393 

When we resolve paths in expressions (either path expressions or paths in struct expressions), there's a need of projection normalization, which `TyLoweringContext` cannot do on its own. We've been properly applying normalization for paths in struct expressions without type anchor, but not for others:

```rust
enum E {
    S { v: i32 }
    Empty,
}

impl Foo for Bar {
    type Assoc = E;
    fn foo() {
        let _ = Self::Assoc::S { v: 42 };   // path in struct expr without type anchor; we already support this
        let _ = <Self>::Assoc::S { v: 42 }; // path in struct expr with type anchor; resolves with this PR
        let _ = Self::Assoc::Empty;         // path expr; resolves with this PR
    }
}
```

With this PR we correctly resolve the whole path, but we need some more tweaks in HIR and/or IDE layers to properly resolve a qualifier (prefix) of such paths and provide IDE features that are pointed out in #14393 to be currently broken.

### CODE REVIEW

File: crates/hir-ty/src/infer.rs [hunk 1]
```diff
@@ -905,8 +873,68 @@ impl<'a> InferenceContext<'a> {
...
+                    return self.resolve_variant_on_alias(ty, None, mod_path);
+                };
+
+                let mut remaining_segments = path.segments().skip(remaining_idx);
+
+                // We need to try resolving unresolved segments one by one because each may resolve
+                // to a projection, which `TyLoweringContext` cannot handle on its own.
+                while !remaining_segments.is_empty() {
+                    let resolved_segment = path.segments().get(remaining_idx - 1).unwrap();
+                    let current_segment = remaining_segments.take(1);
```
Comment (HKalbasi) [inline crates/hir-ty/src/infer.rs:888]: There are a bunch of `unwrap` in this function, which I think it is possible to remove by defining some windows like thing for `PathSegments` (or by using `Itertools::tuple_windows`) and a function which goes from a `&PathSegment` to a `&PathSegments` (similar to `slice::from_ref`).

Comment (Veykril) [reply crates/hir-ty/src/infer.rs:888]: we've got these path segment unwraps in more places (our API for paths isn't really that great currently)

Comment (HKalbasi) [reply crates/hir-ty/src/infer.rs:888]: Okay, so we can do it in a follow up.

**Pull request merged**